File: 2359-find-closest-node-to-given-two-nodes/2359-find-closest-node-to-given-two-nodes.py
class Solution:
    def closestMeetingNode(self, edges: List[int], node1: int, node2: int) -> int:
        graph = {}
        for i in range(len(edges)):
            graph[i] = edges[i]
        
        def search(graph, s):
            idx = 0
            seen = {s: idx}
            while True:
                v_next = graph.get(s, -1)
                if v_next in seen:
                    break
                if v_next != -1:
                    idx += 1
                    seen[v_next] = idx
                    s = v_next
                else:
                    break
            return seen
        
        path1 = search(graph, node1)
        path2 = search(graph, node2)
        
        print(path1)
        print(path2)
        
        res = -1
        nodes_shared = set(path1.keys()) & set(path2.keys())
        if not nodes_shared: return res
        
        running_min = float('inf')
        for node in nodes_shared:
            dis = max(path1[node], path2[node])
            if running_min > dis:
                running_min = dis
                res = node
            elif running_min == dis:
                res = min(node, res)
        return res

File: 2359-find-closest-node-to-given-two-nodes/2359-find-closest-node-to-given-two-nodes.cpp
class Solution {
public:
    int closestMeetingNode(vector<int>& edges, int node1, int node2) {
        int res = -1, dis_min = INT_MAX;
        vector<int> memo1(edges.size(), -1), memo2(edges.size(), -1);
        dfs(node1, 0, edges, memo1);
        dfs(node2, 0, edges, memo2);
        
        for (int i = 0; i < edges.size(); ++i) {
            if(min(memo1[i], memo2[i]) >= 0 && max(memo1[i], memo2[i]) < dis_min) {
                dis_min = max(memo1[i], memo2[i]);
                res = i;
            }
        }
        return res;
    }
private:
    void dfs(int node, int dis, vector<int>& edges, vector<int>& memo) {
        while (node != -1 && memo[node] == -1) {
            memo[node] = dis++;
            node = edges[node];
        }
    }
};